### PR TITLE
fix: Adjust ChoiceId indexes when an option is removed.

### PR DESCRIPTION
**Pull request by timarney**
# Summary | Résumé

Fix for related: 
https://github.com/cds-snc/platform-forms-client/issues/4113
https://github.com/cds-snc/platform-forms-client/issues/3969
https://github.com/cds-snc/platform-forms-client/issues/3627
https://github.com/cds-snc/platform-forms-client/issues/4143

Fixes code to use id for `cleanUpRules(String(id), index);` vs passing the index

Adds code to adjust choiceId indexes when an option is removed

Given an array of rules

```json
conditionalRules: [{ choiceId: "1.0" }, { choiceId: "1.1" }, { choiceId: "1.3" }, { choiceId: "1.4" }, { choiceId: "2.0" }],
```

If the `1.3` option is removed we need to re-index the existing choiceRules to account for an option being removed

In this case remove `1.3`

Look for other rules that match the 1.x pattern

If the rules are `below` the option being removed `leave` them alone 

If they are `above` we need to remove `-1` from the current value

i.e. remove `1.3` ... 1.4  is above ... and now becomes 1.3

resulting in

```json
conditionalRules: [{ choiceId: "1.0" }, { choiceId: "1.1" }, { choiceId: "1.3" }, { choiceId: "2.0" }],
```

https://github.com/user-attachments/assets/2d568303-e45e-4e60-b135-038958d5c212

**Test file:**
[show-hide-rules-form-2024-08-01.json](https://github.com/user-attachments/files/16471364/show-hide-rules-form-2024-08-01.json)

### CODE REVIEW

File: lib/formContext.ts
Comment (thiessenp-cds) [inline lib/formContext.ts:484]: Nice

Comment (timarney) [reply lib/formContext.ts:484]: Was hoping to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/difference

But can't yet

**Comment by thiessenp-cds** [comment]
Nice work!

Testing, removing a rule correctly "cascades" updates to the related rules.

Unrelated I think: (future followup PR)
I ran into what may be an issue that I think is outside the scope of this PR. If I setup a rule to go to a page, delete the page, the logic view updates correctly. But the rule with the deleted page, it still tries to load the deleted page and shows a page in a broken state.

https://github.com/user-attachments/assets/a5b734a1-35ec-434b-b0d7-4b2a18584759

**Review by thiessenp-cds — APPROVED** [review]
Nice work! LGTM :)

**Comment by timarney** [comment]
> Unrelated I think: (future followup PR)

Ya --- unrelated for this PR but group rules also likely need the same sort of logic applied as they also use the same indexes.

**Comment by timarney** [comment]
Holding off on merging this until after the Conditional Logic updates go live

**Comment by timarney** [comment]
Will wait for QA testing before merging.

**Comment by timarney** [comment]
Confirmed the same issue exists for groups / next actions https://github.com/cds-snc/platform-forms-client/issues/4169

**Comment by Abi-Nada** [comment]
QA Tested and LGTM, @timarney ! ✅

**Pull request merged**
